# Week 3: miRNA–mRNA Interaction Integration
## Day 1: Loading and inspecting TargetScan data

In [1]:
import pandas as pd

df_mirna = pd.read_csv('../data/human_mature_miRNAs_features.csv')
df_mirna.head()

header                sequence  \
0  >hsa-let-7a-5p MIMAT0000062 Homo sapiens let-7...  UGAGGUAGUAGGUUGUAUAGUU   
1  >hsa-let-7a-3p MIMAT0004481 Homo sapiens let-7...   CUAUACAAUCUACUGUCUUUC   
2  >hsa-let-7a-2-3p MIMAT0010195 Homo sapiens let...  CUGUACAGCCUCCUAGCUUUCC   
3  >hsa-let-7b-5p MIMAT0000063 Homo sapiens let-7...  UGAGGUAGUAGGUUGUGUGGUU   
4  >hsa-let-7b-3p MIMAT0004482 Homo sapiens let-7...  CUAUACAACCUACUGCCUUCCC   

       mature_name     accession  length  A   C   G  U  GC_content     seed  \
0    hsa-let-7a-5p  MIMAT0000062      22  5   0   8  9    0.363636  GAGGUAG   
1    hsa-let-7a-3p  MIMAT0004481      21  5   6   1  9    0.333333  UAUACAA   
2  hsa-let-7a-2-3p  MIMAT0010195      22  3   9   3  7    0.545455  UGUACAG   
3    hsa-let-7b-5p  MIMAT0000063      22  3   0  10  9    0.454545  GAGGUAG   
4    hsa-let-7b-3p  MIMAT0004482      22  5  10   1  6    0.500000  UAUACAA   

   seed_A  seed_C  seed_G  seed_U  seed_GC_content  seed_group_size  
0       2       0       4       1         0.571429               11  
1       4       1       0       2         0.142857                4  
2       2       1       2       2         0.428571                2  
3       2       0       4       1         0.571429               11  
4       4       1       0       2         0.142857                4

In [2]:
df_targets = pd.read_csv('../data/Predicted_Targets_Context_Scores.default_predictions.txt', sep='\t')
df_targets.head()

Gene ID Gene Symbol      Transcript ID  Gene Tax ID  \
0  ENSG00000121410.7        A1BG  ENST00000263100.3         9544   
1  ENSG00000121410.7        A1BG  ENST00000263100.3         9544   
2  ENSG00000121410.7        A1BG  ENST00000263100.3         9598   
3  ENSG00000121410.7        A1BG  ENST00000263100.3         9598   
4  ENSG00000121410.7        A1BG  ENST00000263100.3         9598   

            miRNA  Site Type  UTR_start  UTR end  context++ score  \
0  mml-miR-23a-3p          3        142      149           -0.428   
1  mml-miR-23b-3p          3        142      149           -0.428   
2     ptr-miR-23a          3        143      150           -0.419   
3     ptr-miR-23b          3        143      150           -0.419   
4     ptr-miR-23c          3        143      150           -0.419   

   context++ score percentile  weighted context++ score  \
0                          97                    -0.388   
1                          97                    -0.388   
2                          97                    -0.419   
3                          97                    -0.419   
4                          97                    -0.419   

   weighted context++ score percentile  Predicted relative KD  
0                                   97                    NaN  
1                                   97                    NaN  
2                                   98                    NaN  
3                                   98                    NaN  
4                                   98                    NaN

In [3]:
df_targets.describe()

Gene Tax ID     Site Type     UTR_start       UTR end  \
count  1.397978e+06  1.397978e+06  1.397978e+06  1.397978e+06   
mean   9.927392e+03  2.077118e+00  1.526225e+03  1.532548e+03   
std    9.447125e+02  7.520834e-01  1.834816e+03  1.834819e+03   
min    8.364000e+03 -3.000000e+00  1.500000e+01  2.100000e+01   
25%    9.598000e+03  2.000000e+00  3.590000e+02  3.660000e+02   
50%    9.615000e+03  2.000000e+00  9.360000e+02  9.420000e+02   
75%    1.009000e+04  3.000000e+00  2.034000e+03  2.040000e+03   
max    1.361600e+04  3.000000e+00  2.447300e+04  2.448000e+04   

       context++ score  context++ score percentile  weighted context++ score  \
count     1.397978e+06                1.397978e+06              1.397978e+06   
mean     -2.004300e-01                8.070422e+01             -1.575153e-01   
std       1.402712e-01                1.789274e+01              1.378899e-01   
min      -2.394000e+00                1.000000e+00             -2.394000e+00   
25%      -2.780000e-01                7.400000e+01             -2.360000e-01   
50%      -1.830000e-01                8.600000e+01             -1.320000e-01   
75%      -9.500000e-02                9.400000e+01             -3.800000e-02   
max      -1.000000e-02                9.900000e+01             -0.000000e+00   

       weighted context++ score percentile  Predicted relative KD  
count                         1.397978e+06          225953.000000  
mean                          7.812715e+01              -3.981482  
std                           2.157391e+01               0.988609  
min                           0.000000e+00              -7.184500  
25%                           6.900000e+01              -4.672200  
50%                           8.600000e+01              -4.003400  
75%                           9.400000e+01              -3.298600  
max                           9.900000e+01               0.375800

In [4]:
# Create df with select columns

df_targetscan = df_targets[['miRNA', 'context++ score', 'context++ score percentile']]
df_targetscan = df_targetscan.rename(columns={'context++ score': 'context_score', 'context++ score percentile': 'context_percentile'})
df_targetscan.head()

miRNA  context_score  context_percentile
0  mml-miR-23a-3p         -0.428                  97
1  mml-miR-23b-3p         -0.428                  97
2     ptr-miR-23a         -0.419                  97
3     ptr-miR-23b         -0.419                  97
4     ptr-miR-23c         -0.419                  97

In [5]:
# FIlter for human miRNAs 

df_targetscan = df_targetscan[df_targetscan['miRNA'].str.startswith('hsa-')]
df_targetscan.describe()

context_score  context_percentile
count  228049.000000       228049.000000
mean       -0.187066           81.403137
std         0.134794           18.008824
min        -2.023000            7.000000
25%        -0.263000           75.000000
50%        -0.170000           88.000000
75%        -0.083000           95.000000
max        -0.010000           99.000000

In [6]:
# Control for unique miRNAs

df_targetscan["miRNA"].nunique()

321

In [7]:
df_targetscan.head(10)

miRNA  context_score  context_percentile
5       hsa-miR-23c         -0.434                  97
6    hsa-miR-23b-3p         -0.434                  97
7    hsa-miR-23a-3p         -0.434                  97
100    hsa-miR-7-5p         -0.010                  39
101  hsa-miR-30b-5p         -0.170                  80
102  hsa-miR-30d-5p         -0.170                  82
103  hsa-miR-30e-5p         -0.146                  75
104  hsa-miR-30a-5p         -0.162                  80
105  hsa-miR-30c-5p         -0.170                  80
106  hsa-miR-758-3p         -0.119                  86

In [8]:
import re

# Find miRNA names that don't follow standard pattern
# Standard pattern: ends with '-number', '-3p', '-5p', or '-number+letter' (e.g., -302e, -23c)
standard_pattern = r'-(3p|5p|\d+[a-z]?)$' # define miRNA pattern using regex. 

outliers = [] #create list of potential outliers
for mirna in df_targetscan['miRNA'].unique(): #loop through miRNAs in df_targetscan 
    if not re.search(standard_pattern, mirna): # condition the loop to only include miRNAs that do not follow the defined standard pattern (i.e. outliers)
        outliers.append(mirna)  # add outliers to outliers list

print(f"Found {len(outliers)} outlier miRNA names:") # print outliers count
for outlier in sorted(outliers): # loop through outliers list and sort 
    print(f"  {outlier}") # print miRNA outliers

Found 33 outlier miRNA names:
  hsa-miR-101-3p.1
  hsa-miR-101-3p.2
  hsa-miR-124-3p.1
  hsa-miR-124-3p.2
  hsa-miR-126-3p.1
  hsa-miR-126-3p.2
  hsa-miR-133a-3p.1
  hsa-miR-133a-3p.2
  hsa-miR-140-3p.1
  hsa-miR-140-3p.2
  hsa-miR-142-3p.1
  hsa-miR-142-3p.2
  hsa-miR-183-5p.1
  hsa-miR-183-5p.2
  hsa-miR-203a-3p.1
  hsa-miR-203a-3p.2
  hsa-miR-302c-3p.1
  hsa-miR-302c-3p.2
  hsa-miR-383-5p.1
  hsa-miR-383-5p.2
  hsa-miR-411-5p.1
  hsa-miR-411-5p.2
  hsa-miR-455-3p.1
  hsa-miR-455-3p.2
  hsa-miR-483-3p.1
  hsa-miR-483-3p.2
  hsa-miR-496.1
  hsa-miR-496.2
  hsa-miR-504-5p.1
  hsa-miR-505-3p.1
  hsa-miR-505-3p.2
  hsa-miR-873-5p.1
  hsa-miR-873-5p.2


In [9]:
# Replacing suffixes from miRNA outliers with 'no characters'

df_targetscan['miRNA'] = df_targetscan['miRNA'].str.replace(r'\.\d+$', '', regex=True)
df_targetscan['miRNA'].value_counts() # sanity check to see if replace function worked. Expect count to be around 304

miRNA
hsa-miR-124-3p     3637
hsa-miR-101-3p     2490
hsa-miR-325-3p     2475
hsa-miR-203a-3p    2404
hsa-miR-424-5p     1939
                   ... 
hsa-miR-450a-5p      10
hsa-miR-151a-5p      10
hsa-miR-551a          8
hsa-miR-551b-3p       8
hsa-miR-412-5p        5
Name: count, Length: 305, dtype: int64

# We are seeing a new miRNA list of 305 unique miRNAs. I expected 304 based on the outliers list, which includes 17 unique miRNAs. Most likely, one of these outlier miRNAs are already in the original list, so this would just add to an already existing count and not a new unique one.

In [10]:
# Next, we need to extract the miRNAs from the miRBase and targetscan dataframes

mb_mirnas = set(df_mirna['mature_name'].unique())
ts_mirnas = set(df_targetscan['miRNA'].unique())

In [11]:
# Check lengths of each set to compare

print(len(mb_mirnas & ts_mirnas)) # unique to each set
print(len(mb_mirnas)) # unique to miRBase only
print(len(ts_mirnas)) # unique to targetscan only

296
2656
305


## Above check shows us what we knew before, that miRBase has 2656 unique miRNAs and targetscan 305. It also gives us a new insight, namely that the two sets share 296 unique miRNAs

In [12]:
# inspect mismatches

sorted(list(ts_mirnas - mb_mirnas))

['hsa-miR-137',
 'hsa-miR-147b',
 'hsa-miR-190b',
 'hsa-miR-217',
 'hsa-miR-320a',
 'hsa-miR-325-3p',
 'hsa-miR-375',
 'hsa-miR-520b',
 'hsa-miR-520e']

## Nine miRNAs in targetscan do not figure in the miRBAse list of miRNas. All but one of these miRNAs are named with mature -3p or -5p suffixes, which explains that. The last one, -325-3p does not exist in miRBase, however, it does as -325. This miRNA is a precursor and gives the name for the only registered mature miRNA, which comes off the 5-end of that precursor. -3p is mentioned in some papers, and has the sequence "UUUAUUGAGGACCUCCUAUCAA". This corresponds with a sequence section at further down towards the 3'end of the -325 precursor

In [13]:
# hsa-miR-325-3p is not denoted in miRBase, however, hsa-miR-325 is. Checking if the theoretical seed and sequence for 325-3p exist 
# in the miRBase list in case it has another name

df_mirna[
    (df_mirna["seed"] == "UUUAUUG") &
    (df_mirna["sequence"] == "UUUAUUGAGGACCUCCUAUCAA")
]

# it does not exist


Empty DataFrame
Columns: [header, sequence, mature_name, accession, length, A, C, G, U, GC_content, seed, seed_A, seed_C, seed_G, seed_U, seed_GC_content, seed_group_size]
Index: []

In [14]:
# We are going to stick to a conservative approach when merging the dataframes. The 8 miRNAs from targetscan that do not show up in the shared
# list will be excluded. We will keep 325-3p, due to its strong prevalence in the targetscan data. The other 8 are not included simply because
# In this project, we do not want to over-complicate things by including miRNAs where the mature version is uncertain. At least for now.
# We remove 8 selected miRNAs later. First, we merge

df_merged = df_targetscan.merge(
    df_mirna,
    left_on='miRNA',
    right_on='mature_name',
    how='left',
    validate='many_to_one'
)

In [15]:
# Checking overlap count, i.e. how many targetscan rows of miRNAs match the mature names from miRBase df

df_merged["mature_name"].isna().value_counts()

#This tells us that 8267 miRNA rows match, i.e. we are seeing many identical miRNAs from targetscan match a single miRNA mature name from miRBase df 


mature_name
False    219782
True       8267
Name: count, dtype: int64

In [16]:
# Let us confirm that we indeed have 296 unique miRNAs that are shared between both dataframes

df_merged['mature_name'].nunique()

296

In [17]:
# Aggregating targetscan data to each miRNA to create the labels describing miRNA interaction properties
# We include mean and min score++ values, and number of interactions 

df_labels = (
    df_targetscan
    .groupby('miRNA', as_index = False) # keep miRNA column as regular column (not index)
    .agg(
        mean_context_score = ('context_score', 'mean'), # "mean" → apply Series.mean() 
        min_context_score = ('context_score', 'min'), # "min" → apply Series.min()
        n_targets = ('context_score', 'count') # "count" → apply Series.count()
    )
)

In [18]:
print(df_labels)

# mean score = average context score
# min score = strongest predicted context score
# n targets = number of targets per miRNA

              miRNA  mean_context_score  min_context_score  n_targets
0     hsa-let-7a-5p           -0.282952             -0.809       1386
1     hsa-let-7b-5p           -0.281561             -0.809       1386
2     hsa-let-7c-5p           -0.283001             -0.809       1386
3     hsa-let-7d-5p           -0.294442             -0.827       1386
4     hsa-let-7e-5p           -0.282975             -0.809       1386
..              ...                 ...                ...        ...
300   hsa-miR-93-5p           -0.181775             -0.721       1643
301   hsa-miR-96-5p           -0.215096             -0.870       1357
302   hsa-miR-98-5p           -0.283528             -0.817       1386
303  hsa-miR-99a-5p           -0.445283             -0.840         60
304  hsa-miR-99b-5p           -0.438433             -0.836         60

[305 rows x 4 columns]


In [19]:
df_labels.describe()

mean_context_score  min_context_score    n_targets
count          305.000000         305.000000   305.000000
mean            -0.196473          -0.757416   747.701639
std              0.076171           0.214137   581.960943
min             -0.445283          -2.023000     5.000000
25%             -0.233377          -0.815000   276.000000
50%             -0.186694          -0.729000   543.000000
75%             -0.149910          -0.630000  1138.000000
max             -0.035642          -0.337000  3637.000000

In [20]:
# We already see an outlier in df_labels (min score for min_context_score is below -1, which is outside the range (0 to -1))
# Let's QC this columns

df_labels.sort_values('min_context_score').head(10)




miRNA  mean_context_score  min_context_score  n_targets
285     hsa-miR-7-5p           -0.283272             -2.023        629
98   hsa-miR-199a-5p           -0.265469             -1.572        723
100  hsa-miR-199b-5p           -0.265524             -1.572        723
76   hsa-miR-181d-5p           -0.139590             -1.487       1694
74   hsa-miR-181b-5p           -0.139475             -1.487       1694
73   hsa-miR-181a-5p           -0.140022             -1.487       1694
75   hsa-miR-181c-5p           -0.137678             -1.479       1694
186      hsa-miR-346           -0.244299             -1.324        197
62   hsa-miR-151a-3p           -0.226929             -1.273        113
150   hsa-miR-29c-3p           -0.262099             -1.239       1463

In [21]:
# QC of labels dataframe

print(df_labels['min_context_score'].dtype)
print(df_labels['mean_context_score'].dtype)
print(df_labels[['mean_context_score', 'min_context_score']].isna().sum())

float64
float64
mean_context_score    0
min_context_score     0
dtype: int64


In [27]:
# Let us remove 8 selected miRNAs from df_labels as discussed above

precursor_only = ['hsa-miR-137',
 'hsa-miR-147b',
 'hsa-miR-190b',
 'hsa-miR-217',
 'hsa-miR-320a',
 'hsa-miR-375',
 'hsa-miR-520b',
 'hsa-miR-520e'
 ]

df_labels = df_labels[~df_labels['miRNA'].isin(precursor_only)]
print(df_labels)

# Confirm the expected 297 rows of unique miRNAs

print(f'There are {df_labels['miRNA'].nunique()} unique miRNAs in df_labels')

              miRNA  mean_context_score  min_context_score  n_targets
0     hsa-let-7a-5p           -0.282952             -0.809       1386
1     hsa-let-7b-5p           -0.281561             -0.809       1386
2     hsa-let-7c-5p           -0.283001             -0.809       1386
3     hsa-let-7d-5p           -0.294442             -0.827       1386
4     hsa-let-7e-5p           -0.282975             -0.809       1386
..              ...                 ...                ...        ...
300   hsa-miR-93-5p           -0.181775             -0.721       1643
301   hsa-miR-96-5p           -0.215096             -0.870       1357
302   hsa-miR-98-5p           -0.283528             -0.817       1386
303  hsa-miR-99a-5p           -0.445283             -0.840         60
304  hsa-miR-99b-5p           -0.438433             -0.836         60

[297 rows x 4 columns]
There are 297 unique miRNAs in df_labels


In [31]:
# COnfirm presence of specially selected miRNA, hsa-miR-325-3p

'hsa-miR-325-3p' in df_labels['miRNA'].values

True

In [23]:
# Let's merge df_labels with selected features from df_mirna
# First, select features

selected_mirna_features = [
    'mature_name',
    'sequence',
    'length',
    'GC_content',
    'seed',
    'seed_GC_content',
    'seed_group_size'
]

df_mirna_features = df_mirna[selected_mirna_features]

df_mirna_features.head()

mature_name                sequence  length  GC_content     seed  \
0    hsa-let-7a-5p  UGAGGUAGUAGGUUGUAUAGUU      22    0.363636  GAGGUAG   
1    hsa-let-7a-3p   CUAUACAAUCUACUGUCUUUC      21    0.333333  UAUACAA   
2  hsa-let-7a-2-3p  CUGUACAGCCUCCUAGCUUUCC      22    0.545455  UGUACAG   
3    hsa-let-7b-5p  UGAGGUAGUAGGUUGUGUGGUU      22    0.454545  GAGGUAG   
4    hsa-let-7b-3p  CUAUACAACCUACUGCCUUCCC      22    0.500000  UAUACAA   

   seed_GC_content  seed_group_size  
0         0.571429               11  
1         0.142857                4  
2         0.428571                2  
3         0.571429               11  
4         0.142857                4

In [33]:
# Merging miRNA features with labels

df_final = df_labels.merge( # df_labels define main row we want to keep 
    df_mirna_features, # this is the table we want to merge with
    left_on='miRNA', # the column identifier we match on 
    right_on='mature_name', # the equivalent colum to 'miRNA'
    how='left', # keep all rows from df_labels (important), add miRNA feature columns where there is a match, and if no match then NaN 
    validate='one_to_one' # safety check to control that each miRNA in df_label must match at most one in df_mirna_feature (protect from silent duplication)
)

df_final.head()

miRNA  mean_context_score  min_context_score  n_targets  \
0  hsa-let-7a-5p           -0.282952             -0.809       1386   
1  hsa-let-7b-5p           -0.281561             -0.809       1386   
2  hsa-let-7c-5p           -0.283001             -0.809       1386   
3  hsa-let-7d-5p           -0.294442             -0.827       1386   
4  hsa-let-7e-5p           -0.282975             -0.809       1386   

     mature_name                sequence  length  GC_content     seed  \
0  hsa-let-7a-5p  UGAGGUAGUAGGUUGUAUAGUU    22.0    0.363636  GAGGUAG   
1  hsa-let-7b-5p  UGAGGUAGUAGGUUGUGUGGUU    22.0    0.454545  GAGGUAG   
2  hsa-let-7c-5p  UGAGGUAGUAGGUUGUAUGGUU    22.0    0.409091  GAGGUAG   
3  hsa-let-7d-5p  AGAGGUAGUAGGUUGCAUAGUU    22.0    0.409091  GAGGUAG   
4  hsa-let-7e-5p  UGAGGUAGGAGGUUGUAUAGUU    22.0    0.409091  GAGGUAG   

   seed_GC_content  seed_group_size  
0         0.571429             11.0  
1         0.571429             11.0  
2         0.571429             11.0  
3         0.571429             11.0  
4         0.571429             11.0

In [34]:
# Some QC

print(df_final.shape)
print(df_final.describe())

(297, 11)
       mean_context_score  min_context_score    n_targets      length  \
count          297.000000         297.000000   297.000000  296.000000   
mean            -0.196839          -0.757104   748.340067   22.000000   
std              0.076509           0.213586   584.172343    0.859641   
min             -0.445283          -2.023000     5.000000   17.000000   
25%             -0.233382          -0.810000   276.000000   22.000000   
50%             -0.187040          -0.729000   543.000000   22.000000   
75%             -0.149910          -0.630000  1138.000000   22.000000   
max             -0.035642          -0.394000  3637.000000   25.000000   

       GC_content  seed_GC_content  seed_group_size  
count  296.000000       296.000000       296.000000  
mean     0.461982         0.478764         2.918919  
std      0.097591         0.158569         2.878355  
min      0.227273         0.000000         1.000000  
25%      0.391304         0.428571         1.000000  
50%     

In [35]:
df_final = df_final.drop(columns=['mature_name'])
df_final.columns

Index(['miRNA', 'mean_context_score', 'min_context_score', 'n_targets',
       'sequence', 'length', 'GC_content', 'seed', 'seed_GC_content',
       'seed_group_size'],
      dtype='object')

## We ahve integrated targetscan df with mirna df, resulting in a table of 10 columns, including labels describing miRNA interaction strength
## as well as various miRNA features that describe the intrinsic properties of each of the 297 included miRNAs